common/src/main/java/com/github/khanshoaib3/minecraft_access/features/narrator_menu/NarratorMenu.java

```diff
@@ -18,26 +22,123 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.biome.Biome;
 
+import java.util.Arrays;
+import java.util.stream.Stream;
+
 /**
  * Opens a menu when F4 button is pressed (configurable) with few helpful options.
  */
 public class NarratorMenu {
     private static MinecraftClient minecraftClient;
+    private static boolean isMenuKeyPressedPreviousTick = false;
+    private static boolean isHotKeyPressedPreviousTick = false;
+    private static boolean isHotKeySwitchedPreviousTick = false;
+    private int hotKeyFunctionIndex = 0;
+    private static final Boolean[] LAST_RUN_HAS_DONE_FLAG = new Boolean[10];
+
+    static {
+        Arrays.fill(LAST_RUN_HAS_DONE_FLAG, Boolean.TRUE);
+    }
+
+    /**
+     * Should be same order as NarratorMenuGUI.init()
+     */
+    private static final MenuFunction[] MENU_FUNCTIONS = new MenuFunction[]{
```

Comment:
@boholder We sohuld also add keypad/numpad keys here.

Reply:
Done